hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala

```diff
@@ -0,0 +1,263 @@
+ * this factory then creates the actual committer implementation.
+ *
+ * @constructor Instantiate. dynamic partition overwrite is not supported,
+ *              so that committers for stores which do not support rename
+ *              will not get confused.
+ * @param jobId                     job
+ * @param dest                      destination
+ * @param dynamicPartitionOverwrite does the caller want support for dynamic
+ *                                  partition overwrite. If so, it will be
+ *                                  refused.
+ * @throws IOException when an unsupported dynamicPartitionOverwrite option is supplied.
+ */
+class PathOutputCommitProtocol(
+    jobId: String,
+    dest: String,
+    dynamicPartitionOverwrite: Boolean = false)
+  extends HadoopMapReduceCommitProtocol(jobId, dest, false) with Serializable {
+
+  if (dynamicPartitionOverwrite) {
+    // until there's explicit extensions to the PathOutputCommitProtocols
+    // to support the spark mechanism, it's left to the individual committer
+    // choice to handle partitioning.
+    throw new IOException(PathOutputCommitProtocol.UNSUPPORTED)
+  }
+
+  /** The committer created. */
+  @transient private var committer: PathOutputCommitter = _
+
+  require(dest != null, "Null destination specified")
+
+  private[cloud] val destination: String = dest
+
+  /** The destination path. This is serializable in Hadoop 3. */
+  private[cloud] val destPath: Path = new Path(destination)
+
+  logDebug(s"Instantiated committer with job ID=$jobId;" +
+    s" destination=$destPath;" +
+    s" dynamicPartitionOverwrite=$dynamicPartitionOverwrite")
+
+  import PathOutputCommitProtocol._
+
+  /**
+   * Set up the committer.
+   * This creates it by talking directly to the Hadoop factories, instead
+   * of the V1 `mapred.FileOutputFormat` methods.
+   * @param context task attempt
+   * @return the committer to use. This will always be a subclass of
+   *         [[PathOutputCommitter]].
+   */
+  override protected def setupCommitter(context: TaskAttemptContext): PathOutputCommitter = {
+
+    logDebug(s"Setting up committer for path $destination")
+    committer = PathOutputCommitterFactory.createCommitter(destPath, context)
+
+    // Special feature to force out the FileOutputCommitter, so as to guarantee
+    // that the binding is working properly.
+    val rejectFileOutput = context.getConfiguration
+      .getBoolean(REJECT_FILE_OUTPUT, REJECT_FILE_OUTPUT_DEFVAL)
+    if (rejectFileOutput && committer.isInstanceOf[FileOutputCommitter]) {
+      // the output format returned a file output format committer, which
+      // is exactly what we do not want. So switch back to the factory.
+      val factory = PathOutputCommitterFactory.getCommitterFactory(
+        destPath,
+        context.getConfiguration)
+      logDebug(s"Using committer factory $factory")
+      committer = factory.createOutputCommitter(destPath, context)
```

Comment:
what's the situation in which this does something different than `PathOutputCommitterFactory.createCommitter`?  Looks like it should be the same? https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/PathOutputCommitterFactory.java#L197

Reply:
It's more sometimes you want the option to say "really, really reject the classic"; a guard against that default actually happening.

The MR committer factories get to choose their own committers; currently the S3A one defaults to (still) using the rename one as switching over to a newer one. I think with HADOOP-16357 we've got the staging committer consistent with what MR expects w.r.t whether or not the destination directory is empty on startup& so for MR and spark work you should be able to switch over to it without worries. A little option here lets you fail fast on job launch to make sure that your job really has done that.

without that, you only get to find out that you are still using it in the good case (it takes a lot longer and you find that `_SUCCESS` is a 0-byte file), or the bad case (it takes a lot longer, you find that `_SUCCESS` is a 0-byte file and because of failures during task commit, your output has been silently corrupted) (*)

(*) I am of the believe that failure condition holds for FileOutputCommitter v2 and any store, it's just that for S3 and other slow-to-rename stores the window of vulnerability is a lot longer, and, without heartbeats coming back, there's a risk of the rename actually triggering a timeout. Never heard of that with Spark, but it has been known to happen with Distcp of a many GB file